main/src/library/RedBlackSet.flix

```diff
@@ -0,0 +1,296 @@
+
+    ///
+    /// Returns true if and only if every element in `xs` appears in `ys`.
+    ///
+    pub def isSubsetOf(xs: RedBlackSet[a], ys: RedBlackSet[a]): Bool = forall(x -> memberOf(x, ys), xs)
+
+    ///
+    /// Returns true if and only if every element in `xs` appears in `ys` and `xs != ys`.
+    ///
+    pub def isProperSubsetOf(xs: RedBlackSet[a], ys: RedBlackSet[a]): Bool = if (isSubsetOf(xs, ys)) size(xs) != size(ys) else false
+
+    ///
+    /// Alias for `findLeft`.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def find(f: a -> Bool, xs: RedBlackSet[a]): Option[a] = findLeft(f, xs)
+
+    ///
+    /// Optionally returns the first element of `xs` that satisfies the predicate `f` when searching from left to right.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def findLeft(f: a -> Bool, xs: RedBlackSet[a]): Option[a] =
+        let result = foldLeft((acc, x) -> if (f(x)) insert(x, acc) else acc, empty(), xs);
+            if (isEmpty(result)) None
+            else
+                let RedBlackSet(es) = result;
+                RedBlackTree.leftMost(es)
+
+    ///
+    /// Optionally returns the first element of `xs` that satisfies the predicate `f` when searching from right to left.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def findRight(f: a -> Bool, xs: RedBlackSet[a]): Option[a] =
+        let result = foldLeft((acc, x) -> if (f(x)) insert(x, acc) else acc, empty(), xs);
+            if (isEmpty(result)) None
+            else
+                let RedBlackSet(es) = result;
+                RedBlackTree.rightMost(es)
+
+    ///
+    /// Alias for `foldLeft`.
+    ///
+    pub def fold(f: (b, a) -> b & e, s: b, xs: RedBlackSet[a]): b & e = foldLeft(f, s, xs)
+
+    ///
+    /// Applies `f` to a start value `s` and all elements in `xs` going from left to right.
+    ///
+    /// That is, the result is of the form: `f(...f(f(s, x1), x2)..., xn)`.
+    ///
+    pub def foldLeft(f: (b, a) -> b & e, s: b, xs: RedBlackSet[a]): b & e =
+        let RedBlackSet(es) = xs;
+            RedBlackTree.foldLeft(f, s, es)
+
+    ///
+    /// Applies `f` to a start value `s` and all elements in `xs` going from right to left.
+    ///
+    /// That is, the result is of the form: `f(x1, ...f(xn-1, f(xn, s))...)`.
+    ///
+    pub def foldRight(f: (b, a) -> b & e, s: b, xs: RedBlackSet[a]): b & e =
+        let RedBlackSet(es) = xs;
+            RedBlackTree.foldRight(f, s, es)
+
+    ///
+    /// Alias for `reduceLeft`.
+    ///
+    pub def reduce(f: (a, a) -> a & e, xs: RedBlackSet[a]): Option[a] & e = reduceLeft(f, xs)
+
+    ///
+    /// Applies `f` to all elements in `xs` going from left to right until a single value `v` is obtained.  Returns `Some(v)`.
+    /// That is, the result is of the form: `Some(f(...f(f(x1, x2), x3)..., xn))`
+    /// Returns `None` if `xs` is the empty set.
+    ///
+    /// TODO: Fix, how to get first value as initial?
```

Comment:
As you comment, reduce cannot readily be expressed in terms of fold, because fold requires an accumulator. Instead, you should patmatch on tree: If it is empty you return None otherwise you have an element and then you can start to reduce.